### PR TITLE
Fix #237 and #256: handle subscription errors

### LINKED ISSUE #237: Report protocol errors that occur after subscriptions to Events and Property updates
see discussions in https://github.com/eclipse/thingweb.node-wot/pull/232#issuecomment-638042931

### PR DESCRIPTION
**Pull request by zolkis**
Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

***
<a href="https://pr-preview.s3.amazonaws.com/zolkis/wot-scripting-api/pull/264.html" title="Last updated on Sep 23, 2020, 9:52 AM UTC (01e3475)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-scripting-api/264/7a59934...zolkis:01e3475.html" title="Last updated on Sep 23, 2020, 9:52 AM UTC (01e3475)">Diff</a>

### CODE REVIEW

**Comment by zolkis** [comment]
Notes:
- Added Subscription interface
- the stop() method name was selected to be aligned with start() and stop() from similar interfaces (e.g. in Generic Sensors)
- also, the name `stop()` makes it possible to add a `start()` method if we want to support temporary stop/start functionality for subscriptions.

Now since `stop()` replaces unsubscribeEvent() and unobserveProperty(), it has to inherit passing InteractionOptions and needs to return a Promise.

**Comment by zolkis** [comment]
Also, factored out the selecting the `unsubscribe` Form for a given `subscribe` Form, which is marked non-normative.
Help is welcome on improving it.

**Review by relu91 — COMMENTED** [review]
Thanks @zolkis. I left some comments they are not really changes but suggestions. Feel free to ignore them . 

On the other hand, now that I am seeing this new API defined I think we are trying too hard to merge two different API styles. The `Subscription` objects feels like EventTarget but it is not, since the event callbacks are passed to `subscribeEvent`/`observeEvent` method.  On the other hand, we'd like to have a promise style for starting/stopping the underlying process. I know that I voted for this style, but it does not feel right. Futhermore, this paragraph from [Web Platform Design Principles](https://w3ctag.github.io/design-principles/#dont-invent-event-like) struck me:

```
For recurring events, it could be convenient to create a custom pair of APIs to "register"/"unregister", "subscribe"/"unsubscribe", etc., 
that take a callback and get invoked multiple times until paired cancel API is called.

Instead, use the existing event registration pattern and separate API controls to start/stop the underlying process (since event listeners should not have side-effects).
```
Since you are already suggesting that in the future we might add also the `start`method, we could go directly with an API similar to GenericSensor API. This means, that Subscription has the start method and the `subscribeEvent`/`observeProperty` will not return a Promise. What do you think?

File: index.html [hunk 1]
```diff
@@ -1463,82 +1435,75 @@ <h3>The <dfn>observeProperty()</dfn> method</h3>
...
-            If the request fails, reject |promise| with the error received from the <a>Protocol Bindings</a> and abort these steps.
-          </li>
           <li>
-            Otherwise resolve |promise|.
-          </li>
-          <li>
-            If the underlying platform receives further notifications for this subscription with new <a>Property</a> value, implementations SHOULD
-            silently suppress them.
+            Whenever the underlying platform receives an error notification for
+            this subscription with an |error:Error|, run the following sub-steps:
```
[the remark this answers is not in the document]
Comment (zolkis) [reply index.html:1497]: We could just say

> Whenever the underlying platform detects errors (i.e. connection errors) for this subscription

File: index.html [hunk 2]
```diff
@@ -1616,77 +1581,304 @@ <h3>The <dfn>subscribeEvent()</dfn> method</h3>
...
+              <li>
+                Let |subscription|'s [[\form]] be the <a>Form</a> associated with
+                |formIndex| in [[\interaction]]'s |forms| array if |option|'s
+                |formIndex| is defined, otherwise let [[\form]] be the first
+                <a>Form</a> in [[\interaction]]'s |forms| array whose |op| is
+                `"subscribeevent"`.
+              </li>
+              <li>
+                If |subscription|'s [[\form]] is failure, reject |promise| with a
+                {{SyntaxError}} and abort these steps.
```
[the remark this answers is not in the document]
Comment (zolkis) [reply index.html:1610]: Also, since it is specifically informative, implementations are free to use a different algorithm.

**Comment by danielpeintner** [comment]
> On the other hand, now that I am seeing this new API defined I think we are trying too hard to merge two different API styles. The `Subscription` objects feels like EventTarget but it is not, since the event callbacks are passed to `subscribeEvent`/`observeEvent` method. On the other hand, we'd like to have a promise style for starting/stopping the underlying process. I know that I voted for this style, but it does not feel right. Futhermore, this paragraph from [Web Platform Design Principles](https://w3ctag.github.io/design-principles/#dont-invent-event-like) struck me:

@relu91 can you sketch the WebIDL how *you* would like to see it so that we have a better idea how it might look like... thanks!

File: index.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (zolkis) [reply index.html:1610]: About the array, it permits a single pass over an unordered set. Keep in mind that contributions for match level could come from different sources. While this algorithm could be optimized in that regard, I deliberately kept it this way to draw attention to this aspect (order of Forms should not affect the outcome of the search). Implementations could choose their algorithm anyway. 

Also, feel free to propose an alternative algorithm, there probably is a better one, this was just a quick one.

**Comment by zolkis** [comment]
> On the other hand, now that I am seeing this new API defined I think we are trying too hard to merge two different API styles. The Subscription objects feels like EventTarget but it is not, since the event callbacks are passed to subscribeEvent/observeEvent method. On the other hand, we'd like to have a promise style for starting/stopping the underlying process. I know that I voted for this style, but it does not feel right. 

I think there are legitimate uses of a subscribe mechanism, see for instance [the Cookie Store API](https://wicg.github.io/cookie-store/) which uses both events and subscribe (with Promise).

I have evaluated the option with returning the Subscription object that has a start() method, in 2 scenarios:
- with Promise, and start() is invoked automatically by subscribe()
- without Promise, one should explicitly start() the subscription.

However, the latter doesn't play well with the semantics of a subscribe() function.

So either I would leave it as it is now, or include a start() method, but let the subscribe() invoke it, returning start()'s Promise.
But it's a complication, since start() should be `void` if we follow the Generic Sensor design (but that one has events so it's fine its start() is void). In our case, start() would need to return a Promise.

So in the end I think this proposal is the minimal one.

**Review by danielpeintner — APPROVED** [review]
I  am ok with the current WebIDL changes. 

More/other improvements can come later...

**Pull request merged**